src/lib/ldap/connection.c

```diff
@@ -549,3 +566,470 @@ int fr_ldap_connection_timeout_reset(fr_ldap_connection_t const *c)
+			 *	If we are chasing a referral, referral_url will be populated and may
+			 *	have a base dn or scope to override the original query
+			 */
+			status = fr_ldap_search_async(&query->msgid, query->request, &ldap_conn,
+						      (referral_url && referral_url->lud_dn) ?
+						      	referral_url->lud_dn : query->dn,
+						      (referral_url && referral_url->lud_scope) ?
+					      		referral_url->lud_scope : query->search.scope,
+					      	      query->search.filter, query->search.attrs,
+						      query->serverctrls, query->clientctrls);
+			break;
+
+		case LDAP_REQUEST_MODIFY:
+			/*
+			 *	This query is an LDAP modification
+			 */
+			status = fr_ldap_modify_async(&query->msgid, query->request, &ldap_conn, query->dn, query->mods, query->serverctrls, NULL);
+			break;
+
+		default:
+			ERROR("Invalid LDAP query for trunk connection");
+			goto error;
+
+		}
+
+		if (status != LDAP_PROC_SUCCESS) goto error;
+
+		/*
+		 *	Record which connection was used for this query
+		 *	- results processing often needs access to an LDAP handle
+		 */
+		query->ldap_conn = ldap_conn;
+
+		/*
+		 *	Add the query to the tree of pending queries for this trunk
+		 */
+		fr_rb_insert(query->ldap_conn->queries, query);
+
+		fr_trunk_request_signal_sent(treq);
+	}
+
+}
+
+/** Read LDAP responses
+ *
+ * Responses from the LDAP server will cause the fd to become readable and trigger this
+ * callback.  Most LDAP search responses have multiple messages in their response - we
+ * only gather those which are complete before either following a referral or passing
+ * the head of the resulting chain of messages back.
+ *
+ * @param[in] tconn	Trunk connection associated with these results.
+ * @param[in] conn	Connection handle for these results.
+ * @param[in] uctx	Thread specific trunk structure - contains tree of pending queries.
+ */
+static void ldap_trunk_request_demux(UNUSED fr_trunk_connection_t *tconn, fr_connection_t *conn, void *uctx)
+{
+	fr_ldap_connection_t	*ldap_conn = talloc_get_type_abort(conn->h, fr_ldap_connection_t);
+	fr_ldap_thread_trunk_t	*t = talloc_get_type_abort(uctx, fr_ldap_thread_trunk_t);
+
+	int 			ret = 0, msgtype;
+	struct timeval		poll = { 0, 10 };
+	LDAPMessage		*result = NULL;
+	fr_ldap_rcode_t		rcode;
+	fr_ldap_query_t		find = { .msgid = -1 }, *query = NULL;
+	request_t		*request;
+
+next_result:
+	/*
+	 *	Look for any results for which we have the complete result message
+	 *	ldap_result will return a pointer to a chain of messages.
+	 */
+	ret = ldap_result(ldap_conn->handle, LDAP_RES_ANY, LDAP_MSG_ALL, &poll, &result);
+
+	switch (ret) {
+	case 0:
+		return;
+
+	case -1:
+		rcode = fr_ldap_error_check(NULL, ldap_conn, NULL, NULL);
+		if (rcode == LDAP_PROC_BAD_CONN) ERROR("Bad LDAP connection");
+		return;
+
+	default:
+		break;
+	}
+
+	find.msgid = ldap_msgid(result);
+	query = fr_rb_find(ldap_conn->queries, &find);
+
+	if (!query) {
+		WARN("Ignoring msgid %i - doesn't match any outstanding queries", find.msgid);
```

Comment:
`WARN("Ignoring msgid %i - doesn't match any outstanding queries (it may have been cancelled)", find.msgid);`